src/DatabaseManager.php

```diff
@@ -170,101 +120,75 @@ public function ensureTenantCanBeCreated(Tenant $tenant): void
+                $chain[] = $item;
+            }
         }
 
-        $this->tenancy->event('database.deleted', $database, $tenant);
+        QueuedTenantDatabaseCreator::withChain($chain)->dispatch($tenant->database()->manager(), $tenant);
     }
 
-    /**
-     * Get the TenantDatabaseManager for a tenant's database connection.
-     *
-     * @param Tenant $tenant
-     * @return TenantDatabaseManager
-     * @throws DatabaseManagerNotRegisteredException
-     */
-    public function getTenantDatabaseManager(Tenant $tenant): TenantDatabaseManager
+    protected function createDatabaseSynchronously(Tenant $tenant, array $afterCreating)
     {
-        $driver = $this->getDriver($this->getBaseConnection($connectionName = $tenant->getConnectionName()));
-
-        $databaseManagers = $this->app['config']['tenancy.database_managers'];
+        $manager = $tenant->database()->manager();
+        $manager->createDatabase($tenant->database()->getName());
 
-        if (! array_key_exists($driver, $databaseManagers)) {
-            throw new DatabaseManagerNotRegisteredException($driver);
+        if ($manager instanceof ManagesDatabaseUsers) {
+            $manager->createUser($tenant->database());
         }
 
-        $databaseManager = $this->app[$databaseManagers[$driver]];
-
-        if ($connectionName !== 'tenant' && $databaseManager instanceof CanSetConnection) {
-            $databaseManager->setConnection($connectionName);
+        foreach ($afterCreating as $item) {
+            if (is_object($item) && ! $item instanceof Closure) {
+                $item->handle($tenant);
+            } else {
+                $item($tenant);
+            }
         }
-
-        return $databaseManager;
     }
 
     /**
-     * What key on the connection config should be used to separate tenants.
+     * Delete a tenant's database.
      *
-     * @param string $connectionName
-     * @return string
+     * @throws DatabaseManagerNotRegisteredException
      */
-    public function separateBy(string $connectionName): string
+    public function deleteDatabase(Tenant $tenant)
     {
-        if ($this->getDriver($this->getBaseConnection($connectionName)) === 'pgsql'
-            && $this->app['config']['tenancy.database.separate_by'] === 'schema') {
-            return 'schema';
+        $database = $tenant->database()->getName();
+        $manager = $tenant->database()->manager();
+
+        $this->tenancy->event('database.deleting', $database, $tenant);
+
+        if ($this->app['config']['tenancy.queue_database_deletion'] ?? false) {
+            QueuedTenantDatabaseDeleter::dispatch($manager, $tenant);
+        } else {
+            $manager->deleteDatabase($database);
+            if ($manager instanceof ManagesDatabaseUsers) {
+                $manager->deleteUser($tenant->database());
+            }
         }
 
-        return 'database';
+        $this->tenancy->event('database.deleted', $database, $tenant);
```

Comment:
Don’t you need to wait for the queued job to be completed before dispatching this event?

Reply:
That's a fair point. This got me thinking that I can maybe get rid of this clumsy `afterCreating` logic, and just make people use the `created` event for stuff that should be executed after the database is actually created? I'd trigger this event inside the queued job, and it would even make more sense because now it's triggered before the job is executed (after it's dispatched):

https://github.com/stancl/tenancy/blob/a39d961cc3e21721fe99c6410cccb9c213675ed4/src/DatabaseManager.php#L129-L136

The logic currently looks like this:

https://github.com/stancl/tenancy/blob/a39d961cc3e21721fe99c6410cccb9c213675ed4/src/DatabaseManager.php#L138-L164

So instead of dealing with queue chains and other complex stuff like that, I'm thinking that I could just dispatch this event inside the `QueuedTenantDatabaseCreater` job.

And of course, the same would apply for deleting DBs.

Reply:
Will do this as part of another PR, to avoid making this one too large.

Reply:
Great idea. Stay green with new test and merge in 👍🏻